### PR TITLE
Test VHD content for Linux

### PR DESCRIPTION
**Pull request by adilliadil**
This change helps us to test the vhd content each time we do a build.

Tested for:
- [x] docker
- [x] containerd

### CODE REVIEW

File: vhdbuilder/packer/test/run-test.sh [hunk 1]
```diff
@@ -10,108 +11,123 @@ RESOURCE_GROUP_NAME="$TEST_RESOURCE_PREFIX-$(date +%s)"
...
+  else
+    az group delete --name $RESOURCE_GROUP_NAME --yes --no-wait
+  fi
 }
 trap cleanup EXIT
 
 DISK_NAME="${TEST_RESOURCE_PREFIX}-disk"
 VM_NAME="${TEST_RESOURCE_PREFIX}-vm"
 
+#OS_DISK_URI="https://aksimages1610268313.blob.core.windows.net/system/Microsoft.Compute/Images/aks-vhds/aks-1610268313-osDisk.fe9a225e-612e-4a15-80a3-2976f38104f4.vhd"
```
Comment (adilliadil) [inline vhdbuilder/packer/test/run-test.sh:26]: remove

File: vhdbuilder/packer/test/run-test.sh [hunk 2]
```diff
@@ -10,108 +11,135 @@ RESOURCE_GROUP_NAME="$TEST_RESOURCE_PREFIX-$(date +%s)"
...
 
-# we have to use `-E` to disable interpretation of backslash escape sequences, for jq cannot process string 
-# with a range of control characters not escaped as shown in the error below:
-#   Invalid string: control characters from U+0000 through U+001F must be escaped
-errMsg=$(echo -E $ret | jq '.value[]  | select(.code == "ComponentStatus/StdErr/succeeded") | .message')
-# a successful errMsg should be '""' after parsed by `jq`
-if [[ $errMsg != \"\" ]]; then
+if [ "$OS_TYPE" == "Linux" ]; then
+  SCRIPT_PATH="$CDIR/$LINUX_SCRIPT_PATH"
+  sleep 900
```
Comment (adilliadil) [inline vhdbuilder/packer/test/run-test.sh:76]: az vm run-command does not work if we don't sleep here, I don't exactly know why it's taking so much time for the VM to be "run-command ready", but feel like this may be a series performance issue

File: vhdbuilder/packer/test/linux-vhd-content-test.sh [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+err() {
+  echo "$1:Error: $2" >>/dev/stderr
+}
+
+string_replace() {
+  echo $1 | sed "s/\*/$2/"
+}
+
+filesToDownload='
```
Comment (adilliadil) [inline vhdbuilder/packer/test/linux-vhd-content-test.sh:83]: I will add these to a separate file in another PR which will include a Refactor for the install-dependencies.sh file

File: vhdbuilder/packer/install-dependencies.sh [hunk 4]
```diff
@@ -353,41 +353,41 @@ done
 
 # calico images used by AKS
 CALICO_CNI_IMAGES="
-v3.8.9
-v3.8.9.1
+3.8.9
```
Comment (eriksywu) [inline vhdbuilder/packer/install-dependencies.sh:356]: why the change

Comment (adilliadil) [reply vhdbuilder/packer/install-dependencies.sh:356]: this is just for consistencies sake, all the versions don't start with v for the other dependencies, so wanted to be consistent here, nothing serious ;)

File: vhdbuilder/packer/test/linux-vhd-content-test.sh [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+      err $test "Directory ${downloadLocation} does not exist"
+      missingPaths+=("$downloadLocation")
+      continue
+    fi
+
+    for version in ${versions}; do
+      file_Name=$(string_replace $fileName $version)
+      dest="$downloadLocation/${file_Name}"
+
+      if [ ! -s $dest ]; then
```
Comment (mainred) [inline vhdbuilder/packer/test/linux-vhd-content-test.sh:26]: Downloading the file and compare MD5 is time-consuming, but we can compare the size of remote file with the local file. 
`curl -sI $URL | grep -i Content-Length | awk '{print $2}'` will get the file size from the response header without downloading the file

File: vhdbuilder/packer/test/run-test.sh (hunk 2 above)
Comment (mainred) [reply vhdbuilder/packer/test/run-test.sh:76]: Do we need to wait more than 10 minutes after we get a ready VM? I did not find this issue for Windows case. In windows I hit a problem that docker is not ready when try to compare the images.

Comment (adilliadil) [reply vhdbuilder/packer/test/run-test.sh:76]: as of now, this is the case, the run command keeps on retrying until it times out, so that's why I have added an extra 15minutes so it doesn't time out, but I'm investigating this issue why it is happening

File: vhdbuilder/packer/test/linux-vhd-content-test.sh (hunk 5 above)
Comment (adilliadil) [reply vhdbuilder/packer/test/linux-vhd-content-test.sh:26]: sure working on it

**Review by adilliadil — COMMENTED** [review]
working on suggested changes

File: vhdbuilder/packer/test/linux-vhd-content-test.sh (hunk 5 above)
Comment (adilliadil) [reply vhdbuilder/packer/test/linux-vhd-content-test.sh:26]: done

File: vhdbuilder/packer/test/run-test.sh (hunk 2 above)
Comment (adilliadil) [reply vhdbuilder/packer/test/run-test.sh:76]: confirmed that this is not the case if we use a sig image then create VM for any region even for eastus.

Comment (adilliadil) [reply vhdbuilder/packer/test/run-test.sh:76]: still trying to create the very same scenario for another region

Comment (bowang-666) [reply vhdbuilder/packer/test/run-test.sh:76]: is it possible to check the VM status? i.e. polling the status until it is ready

Comment (bowang-666) [reply vhdbuilder/packer/test/run-test.sh:76]: another optimization could be wrap the "az vm run-command" into a retry logic with timeout, check retrycmd_if_failure for example.

**Comment by bowang-666** [comment]
lgtm other than the sleep logic

File: vhdbuilder/packer/test/run-test.sh (hunk 2 above)
Comment (adilliadil) [reply vhdbuilder/packer/test/run-test.sh:76]: tested this for couple of other regions as well, so looks like this is a common pattern. However, this does not happen for VMs created from SIG images.
@bowang-666 VM status is set to succeeded right away so querying the status won't help us. However, we may retry.

Comment (bowang-666) [reply vhdbuilder/packer/test/run-test.sh:76]: Sounds good. Please change the sleep to retry, which will allow it to end dynamically.

Comment (adilliadil) [reply vhdbuilder/packer/test/run-test.sh]: Done

File: vhdbuilder/packer/test/run-test.sh [hunk 6]
```diff
@@ -10,108 +11,137 @@ RESOURCE_GROUP_NAME="$TEST_RESOURCE_PREFIX-$(date +%s)"
...
-    az vm create\
+  # In SIG mode, Windows VM requires admin-username and admin-password to be set,
+  # otherwise 'root' is used by default but not allowed by the Windows Image. See the error image below:
+  # ERROR: This user name 'root' meets the general requirements, but is specifically disallowed for this image. Please try a different value.
+  az vm create \
     --resource-group $RESOURCE_GROUP_NAME \
     --name $VM_NAME \
     --image $IMG_DEF \
     --admin-username $Test_VM_ADMIN_USERNAME \
     --admin-password $TEST_VM_ADMIN_PASSWORD \
```
Comment (yizhang4321) [inline vhdbuilder/packer/test/run-test.sh:53]: --admin-username $Test_VM_ADMIN_USERNAME \
--admin-password $TEST_VM_ADMIN_PASSWORD \

Why the capitalization difference: one is $Test_***, the other is $TEST_***. (I understand this is not new code. Just curious why.)

**Review by yizhang4321 — APPROVED** [review]

File: vhdbuilder/packer/test/run-test.sh (hunk 6 above)
Comment (adilliadil) [reply vhdbuilder/packer/test/run-test.sh:53]: :), yeah that doesn't seem right, fixing it :)

File: .pipelines/.vsts-vhd-builder.yaml [hunk 7]
```diff
@@ -77,6 +73,28 @@ phases:
...
+          -e AZURE_LOCATION=${AZURE_LOCATION} \
+          -e CONTAINER_RUNTIME=${CONTAINER_RUNTIME:-"docker"} \
+          -e UBUNTU_SKU=${UBUNTU_SKU} \
+          -e OS_TYPE="Linux" \
+          -e MODE=$(MODE) \
+          -e VHD_DEBUG=${VHD_DEBUG} \
+          -e SIG_GALLERY_NAME=${SIG_GALLERY_NAME} \
+          -e SIG_IMAGE_NAME=${SIG_IMAGE_NAME} \
+          -e SIG_IMAGE_VERSION=${SIG_IMAGE_VERSION} \
+          ${CONTAINER_IMAGE} make -f packer.mk test-building-vhd
```
Comment (xuto2) [inline .pipelines/.vsts-vhd-builder.yaml:96]: why we don't update .pipelines/templates/.builder-release-template.yaml?

File: vhdbuilder/packer/test/linux-vhd-content-test.sh [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+},
+{
+  "fileName":"v*/bpftrace-tools.tar",
+  "downloadLocation":"/opt/bpftrace/downloads",
+  "downloadURL":"https://upstreamartifacts.azureedge.net/bpftrace",
+  "versions": ["0.9.4"]
+}
+]
+'
+imagesToBePulled='
```
Comment (xuto2) [inline vhdbuilder/packer/test/linux-vhd-content-test.sh:117]: we should also test kubelet/kubectl/kube-proxy
